### ISSUE TITLE
Choose my own domainname and TLS certificate for "managed private repository" making it true as-a-service

### ISSUE THREAD

**Issue by peter-englmaier** [ovh/public-cloud-roadmap#96]
With "managed private registry" we can host goharbor from an address https://*.container-registry.ovh.net/ using OVH's public cloud as-a-service. No other service from the public cloud is needed for that. The address is served by a wild card certificate from OVH.

However, as a software company, we want to deliver our software to our customers using our own domain name. This is not yet possible, but it seems trivial to extend the existing setup to allow setting FQDN and uploading certificates for the service. The IP address must be fixed also (currently this seems to be the case). Using let's-encrypt might be another option.

**Comment by mhurtrel** [comment]
Hi Peter 

I confirm this is a feature we have on our internal roadmap, for the mid term and was missing here.
My current estimation for this feature would be late 2021. I make it part of the public backlog here and edit the title to normalize it.

**Comment by 131** [comment]
Any update here ?

**Comment by mhurtrel** [comment]
Hi @131 I confirm the feature is still on the roadmap, but we decided to move it after the upgrade to a more recent version of Harbor, that will bring a lot of new value. This customdomain feature will come early 2023, sorry for the delay.

**Comment by mhurtrel** [comment]
A small update to confirm we will work on this feature in the first part of 2023 calendar year. A more precise ETA will be share in a few weeks.

**Comment by mhurtrel** [comment]
Confirming we think we will be able to work on this feature this summer

**Comment by YohannHammad** [comment]
Hello @mhurtrel, any news about the precise ETA ?

**Comment by mhurtrel** [comment]
Sorry @YohannHammad we cannot share an ETA yet

**Comment by JeanRessouche** [comment]
Up, pretty sure that you underestimate the number of customers waiting for it.
Had to choose a different host for 5 instances since 2021 due to the missing feature :'(

**Comment by jonchun** [comment]
I know I'm late to the party, but I'll post this here in case anyone else using k8s wants to do something similar. 

TLDR: I gave up on trying to use both the official and the bitnami helm charts for harbor. I also didn't feel like having yet another app to maintain but had the same requirement as everyone else. The below configurations reverse proxy to the OVH managed registry and allow me to use my own name. You could do the same thing with just regular nginx.

Obvious downside here is it will use up bandwidth on wherever you are hosting your reverse proxy -- not an issue for me right now but this may be an issue in the future

Below is TESTED with ingress-nginx + cert-manager
```
apiVersion: v1
kind: Service
metadata:
  name: registry-proxy
  namespace: registry
spec:
  type: ExternalName
  externalName: xxxxx.xx.xx.container-registry.ovh.us
---

apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: registry-proxy
  namespace: registry
  annotations:
    kubernetes.io/ingress.class: "nginx"
    nginx.ingress.kubernetes.io/proxy-body-size: "0"
    nginx.ingress.kubernetes.io/proxy-read-timeout: "60"
    nginx.ingress.kubernetes.io/proxy-send-timeout: "60"
    nginx.ingress.kubernetes.io/proxy-ssl-verify: "true"
    nginx.ingress.kubernetes.io/proxy-buffering: "off"
    nginx.ingress.kubernetes.io/ssl-redirect: "true"
    nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
    nginx.ingress.kubernetes.io/proxy-ssl-protocols: "TLSv1.2 TLSv1.3"
    nginx.ingress.kubernetes.io/upstream-vhost: xxxxx.xx.xx.container-registry.ovh.us
    nginx.ingress.kubernetes.io/proxy-set-headers: |
      proxy_set_header Content-Type $upstream_http_content_type;
      proxy_set_header Accept $http_accept;
    cert-manager.io/issuer-kind: "ClusterIssuer"
    cert-manager.io/cluster-issuer: "letsencrypt-prod"
    cert-manager.io/dns-names: "r.mydomain.com"
spec:
  ingressClassName: nginx
  tls:
  - hosts:
    - r.mydomain.com
  rules:
  - host: r.mydomain.com
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: registry-proxy
            port:
              number: 443
```

Below is UNTESTED. Just asked AI, but from a spot-check it looks pretty good. You should be able to get it to work with minor tweaks.
```
server {
    listen 443 ssl;
    server_name r.mydomain.com;

    ssl_certificate /etc/nginx/secrets/registry-cert.crt;
    ssl_certificate_key /etc/nginx/secrets/registry-cert.key;
    ssl_protocols TLSv1.2 TLSv1.3;
    ssl_ciphers 'ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES128-GCM-SHA256';

    location / {
        proxy_pass https://xxxxx.xx.xx.container-registry.ovh.us;
        proxy_set_header Host $host;
        proxy_set_header Accept $http_accept;
        proxy_set_header Content-Type $upstream_http_content_type;

        proxy_buffering off;
        proxy_ssl_verify on;
        proxy_ssl_protocols TLSv1.2 TLSv1.3;
        proxy_read_timeout 60;
        proxy_send_timeout 60;

        # Allow unlimited body size if needed
        client_max_body_size 0;
    }
}

server {
    listen 80;
    server_name r.mydomain.com;

    # Redirect HTTP to HTTPS
    return 301 https://$host$request_uri;
}
```

Edit: i just thought about this a bit more... surely this is a security risk? I'm guessing the login pages of the Harbor project don't have any sort of "allowed domains" protection? 

I will probably probably not be using my posted solution but will keep it there for anyone else who might choose to do so.

---

@mhurtrel - Any updates on this feature?

**Comment by hrabal-myo** [comment]
Are we having any progress on this matter? It is a bit of a joke to have this request open since 2021 and the only answers posted here are: We are planning to this next year. Do I really have to go somewhere else or is this finally planned in?

**Comment by tux3** [comment]
For us, we looked at the proxy solution but it has serious issues, this really needs to be supported by the cloud provider to work correctly.

If you use `registry.contoso.com` as external name, a proxy can rewrite the requests to `xxx.container-registry.ovh.net`, but it can't rewrite the image name. You can do `docker login registry.contoso.com`, but you cannot do `docker push registry.contoso.com/myproject/myimage`, because Harbor will see you tagged something with an external hostname it doesn't know about, and it will just bail out.

**Comment by antonin-a** [comment]
Hello @tux3 , thank you for your feedback, indeed the proxy solution is not a satisfying workaround. I'm sorry for the delay on this one but I do confirm that custom domain name feature for Managed Private Registry is planed for end of this year. We will keep you updated here.

**Comment by fr34kyn01535** [comment]
Yea I already went for the reverse proxy route for now (initally accidently changing the registry of my nginx-ingress - thus blowing it when it tried to restart :D). This is really a needed feature here, when offering the registry, having to change the name in all of your git ops is quite a hassle, being able to reuse a hostname would also make migration easier.

**Comment by MauritsMonteyne** [comment]
@antonin-a Is this still something we can expect in perhaps early 2026 or have priorities shifted?

**Comment by mdiermaier** [comment]
This is really annoying that such a simple feature is still not available as of 2026. I mean this thread is open for 5 years and nothing happened so far. Please consider implementing. Especially with regards to european data sovereignty this is a MUST HAVE in 2026!

**Comment by arnoschoon** [comment]
We're currently using a proxy-setup as described by @jonchun for AWS ECR, but with data sovereignty in mind we have to choose between self-hosting it completely or use a managed service. A custom domain is bit of a requirement for the migration to be manageable thus for getting the go-ahead from leadership.

Is there anything we can do to speed up delivery of this feature?

**Comment by gdubcs** [comment]
Hi there,
Waiting for this feature as well... Amazing that after 5+ years, not only is it still not implemented, but the lack of communication (or even consideration) is astounding. I guess no one cares about this at OVH.

**Comment by medireport-jean** [comment]
Indeed, and that's certainly not hard to implement...

**Comment by lclc** [comment]
We require this to move our private registry to OVH.

Let us know if/how we can assist.

**Comment by julian-heckmann** [comment]
This is also blocking us from migrating to OVH managed private registry.

**Comment by lclc** [comment]
It says: Status: Prioritized since Sep 26, 2023;; These issues should be addressed within the next 12 months (see https://github.com/orgs/ovh/projects/16/views/1).

@nicolaslacroux maybe you can have a look at this. From his GitHub account it looks like @antonin-a is on leave or so.
